### PR TITLE
Google APIs integration improvements

### PR DESCRIPTION
**Pull request by aerosol**
### Changes

This PR aims to fix several bugs and regressions. Unfortunately due to tight coupling it's not really possible to split this into smaller pieces. Reviewing commits one by one should help a little.

Summary of changes:

1. Referrer drilldown should no longer crash in case of Google API errors, be it invalid grant, insufficient scope or random HTTP errors (sentry: APP-1ZY)

2. Any search console integration errors should be now caught correctly, both in the search console setting and in the react component. React component displays an info visible only to the site owners:

image

This is a small improvement. At first I made an attempt at translating all the possible error codes but that gets messy very quickly - arbitrary lengths and messages don't play well with this small UI window. We don't have a standardised way of translating error codes into human readable error messages in react either, so there's definitely room for future improvement here. However, from there the user can be taken to the search console settings where we do our best to reveal the details (a regression, now tested, was found for the "invalid grant" error too):

image

It is currently my working assumption that the "invalid grant" error can happen any any stage, including Google Token refresh and is always carried with HTTP 400. In that case the following error message is presented to the user:

> Invalid Grant error returned from Google. See here on how to fix it. 

(linking to https://plausible.io/docs/google-search-console-integration#i-get-the-invalid-grant-error)

In case of 401 and 403 errors, we assume this is permanent integration error - either invalid API scope (the user forgot to tick all the boxes) or an invalid token. In this case, the user is presented with a re-link suggestion:

> Your Search Console account hasn't been connected successfully. Please unlink your Google account and try linking it again.

If the above fails, and we're yet experiencing another type of error, a generic message is shown:

> Something went wrong, but looks temporary. If the problem persists, try re-linking your Google account.

3. Non-user facing changes: tests were added and cleaned up a little, as we're slowly moving away from Double mocks in favor of Mox. This helps us finally test any integration failure modes going forward.

### Tests
- [x] Automated tests have been added
- [ ] This PR does not require tests

### Changelog
- [ ] Entry has been added to changelog
- [x] This PR does not make a user-facing change

### Documentation
- [ ] [Docs](https://github.com/plausible/docs) have been updated
- [x] This change does not need a documentation update

### Dark mode
- [x] The UI has been tested both in dark and light mode
- [ ] This PR does not change the UI

### CODE REVIEW

File: lib/plausible/google/http.ex [hunk 1]
```diff
@@ -40,12 +40,12 @@ defmodule Plausible.Google.HTTP do
          {:ok, report} <- convert_to_maps(report) do
       {:ok, {report, token}}
     else
-      {:error, %{reason: %{status: _non_http_200, body: body}}} ->
-        Sentry.Context.set_extra_context(%{google_analytics_response: body})
+
+      {:error, %{reason: %{status: status, body: body} = response}} ->
+        Sentry.Context.set_extra_context(%{ga_response: %{body: body, status: status}})
```
Comment (aerosol) [inline lib/plausible/google/http.ex:45]: cc @vinibrsl - this is https://github.com/plausible/analytics/commit/95aafb477d46857af2a330653bc6f76e027a5329 incorporated

File: lib/plausible_web/templates/site/settings_search_console.html.eex [hunk 2]
```diff
@@ -36,12 +36,24 @@
...
         <% {:error, error} -> %>
-          <p class="text-gray-700 dark:text-gray-300 mt-6">The following error happened when fetching your Google Search Console domains.</p>
+          <p class="text-gray-700 dark:text-gray-300 mt-6">The following error happened when fetching your Google Search Console domains:</p>
 
-          <%= if error == "invalid_grant" do %>
-            <p class="text-red-700 font-medium mt-3"><a href="https://plausible.io/docs/google-search-console-integration#i-get-the-invalid-grant-error">Invalid Grant error returned from Google. <span class="text-indigo-500">See here on how to fix it.</a></p>
-          <% else %>
-            <p class="text-red-700 font-medium mt-3">Something went wrong</p>
+          <%= case error do %>
+            <% "invalid_grant" ->  %>
```
Comment (vinibrsl) [inline lib/plausible_web/templates/site/settings_search_console.html.eex:42]: What do you think of returning atoms instead of strings?

**Review by vinibrsl — APPROVED** [review]
This is a huge improvement! Thanks a lot :)

File: lib/plausible_web/templates/site/settings_search_console.html.eex (hunk 2 above)
Comment (aerosol) [reply lib/plausible_web/templates/site/settings_search_console.html.eex:42]: We can mix them, but the errors dynamically fetched from Google responses are strings after all

**Comment by aerosol** [comment]
> Were you able to test it locally with GA OAuth?

To an extent, I tested some of the failures locally. We should probably put that on staging to make sure there's no regressions.

File: lib/plausible_web/templates/site/settings_search_console.html.eex (hunk 2 above)
Comment (vinibrsl) [reply lib/plausible_web/templates/site/settings_search_console.html.eex:42]: Let's keep it as strings then

Comment (aerosol) [reply lib/plausible_web/templates/site/settings_search_console.html.eex:42]: I think going forward as we clean this up, we could introduce error structs containing error codes and messages for example. But that's for another day.

**Comment by aerosol** [comment]
@vinibrsl @RobertJoonas let's not merge this just yet, I have some more testing to do

File: test/support/test_utils.ex [hunk 3]
```diff
@@ -2,6 +2,34 @@ defmodule Plausible.TestUtils do
...
   alias Plausible.Factory
 
+  defmacro __using__(_) do
+    quote do
+      require Plausible.TestUtils
+      import Plausible.TestUtils
+    end
+  end
+
+  defmacro patch_env(env_key, value) do
```
Comment (ruslandoga) [inline test/support/test_utils.ex:12]: can it be a function?

**Pull request merged**